### PR TITLE
[AzureRT] December release

### PR DESCRIPTION
**Pull request by hyonholee**
- Add "Force" parameter to Remove cmdlets (Breaking change)
- Remove unnecessary (redundant) properties from VM object.
- Add DisplayHint property to VM object to enable Compact and Expand display modes
- Add "Remove-AzureRmVMSecret" cmdlet

## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThrough` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs [hunk 1]
```diff
@@ -42,6 +42,7 @@ public class RemoveAzureRmContainerServiceAgentPoolProfileCommand : Microsoft.Az
         [Parameter(
             Mandatory = true,
             Position = 1,
+            ParameterSetName = "NameParameterSet",
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs]: @hyonholee why was this parameter set added?

File: src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/RemoveAzureRmVmssExtensionCommand.cs [hunk 2]
```diff
@@ -42,9 +42,17 @@ public class RemoveAzureRmVmssExtensionCommand : Microsoft.Azure.Commands.Resour
         [Parameter(
             Mandatory = true,
             Position = 1,
+            ParameterSetName = "NameParameterSet",
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/RemoveAzureRmVmssExtensionCommand.cs]: @hyonholee same comment about parameter set

File: src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/RemoveAzureRmVmssNetworkInterfaceConfigurationCommand.cs [hunk 3]
```diff
@@ -42,9 +42,17 @@ public class RemoveAzureRmVmssNetworkInterfaceConfigurationCommand : Microsoft.A
         [Parameter(
             Mandatory = true,
             Position = 1,
+            ParameterSetName = "NameParameterSet",
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/RemoveAzureRmVmssNetworkInterfaceConfigurationCommand.cs]: @hyonholee same question about parameter set

File: src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs [hunk 4]
```diff
@@ -88,34 +91,7 @@ public string ResourceGroupName
         // Gets or sets the storage profile.
         public StorageProfile StorageProfile { get; set; }
 
-        [JsonIgnore]
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs]: @hyonholee why were these properties removed?

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs [hunk 5]
```diff
@@ -39,9 +39,8 @@ public class RemoveAzureVMDataDiskCommand : Microsoft.Azure.Commands.ResourceMan
 
         [Alias("Name")]
         [Parameter(
-            Mandatory = true,
+            Mandatory = false,
             Position = 1,
-            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs]: @hyonholee why did you remove this attribute? Scripts that get `DataDiskNames` from the pipeline will no longer work

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMNetworkInterfaceCommand.cs [hunk 6]
```diff
@@ -42,9 +42,8 @@ public class RemoveAzureVMNetworkInterfaceCommand : Microsoft.Azure.Commands.Res
 
         [Alias("Id", "NicIds")]
         [Parameter(
-            Mandatory = true,
+            Mandatory = false,
             Position = 1,
-            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMNetworkInterfaceCommand.cs]: @hyonholee same comment about removing this attribute

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMSecretCommand.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Compute
+{
+    [Cmdlet(
+        VerbsCommon.Remove,
+        ProfileNouns.VaultSecretGroup),
+    OutputType(
+        typeof(PSVirtualMachine))]
+    public class RemoveAzureVMSecretCommand : Microsoft.Azure.Commands.ResourceManager.Common.AzureRMCmdlet
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMSecretCommand.cs]: @hyonholee since this cmdlet makes changes, it should implement `SupportsShouldProcess`

**Comment by cormacpayne** [comment]
@hyonholee the build is failing because there is no help for the new cmdlet `Remove-AzureRmVMSecret`, and the same cmdlet does not implement `SupportsShouldProcess`

File: src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs (hunk 4 above)
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs]: This is unnecessary.  This property is just a copy of a nested property, which was used for piping scenario, but now the piping scenario works without these properties.  There was a complain about these redundant property.

File: src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs (hunk 1 above)
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs]: We generated these cmdlets.  For Removing there are can be more than one way to removing (for example, removing with a name and removing with a id), so we added parameter set name for each removing way.

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs (hunk 5 above)
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs]: Pipeline works because I also updated Remove-AzureRmVMDataDisk.  Please look at Test-VirtualMachineProfile test.

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMSecretCommand.cs (hunk 7 above)
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMSecretCommand.cs]: This cmdlet only updates a local VM object and does not actually remove the secret from VM, until Update-AzureRmVM is performed with the updated VM object.  Other Remove cmdlets that change only local object do not have ShouldProcess or Force parameter

File: src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs (hunk 1 above)
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs]: If this is concerned, I will update our generator and remove parameter set when there is only one way to remove.

**Comment by hyonholee** [comment]
I will update the help file.  Thanks for comments.

File: src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs (hunk 1 above)
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService/Config/RemoveAzureRmContainerServiceAgentPoolProfileCommand.cs]: I was just wondering why it was added. As long as the scenario where only the `ContainerService` parameter is provided is covered, this will be fine

File: src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs (hunk 4 above)
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs]: There are numerous cmdlets that output an object of type `PSVirtualMachine`, and existing scripts that use this object will no longer be able to use these properties, which means it's a breaking change.

Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs]: Can we announce the breaking change?  Do you think it is serious?  These properties are not included in the VM response at the top level.  They are nested properties and we added these redundant properties at the top level just for pipeline.

+ @singhkays

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs (hunk 5 above)
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs]: Actually there is an issue with pipeline with Remove-AzureRmVMDataDisk (and other Remove command that removes a property from VM object.)  Since this parameteter (DataDiskNames) have an alias "Name" and VM object has "Name" property for VM Name, in piping scenario (e.g. Get-AzureRmVM | Remove-AzureRmVMDataDisk ), VM name is piped to Remove-AzureRmVMDataDisk as a data disk name, and it tries to remove a data disk named VM name.   Setting ValueFromPipelineByPropertyName to false removes this issue.  Also we remove "DataDiskNames" (which is redundant property) from VM object, DataDiskNames are not piped from VM object to Remove-AzureRmVM.    Remove-AzureRmVM is also modified when DataDiskNames are not provided, it interprets it as removing all existing data disks.  So the command does not change the behavior of Get-AzureRmVM | Remove-AzureRmVM .

File: src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs (hunk 4 above)
Comment (singhkays) [inline src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs]: @hyonholee Not sure this release is setup to be a breaking change. Can we leave this in the object until next time we have a breaking change release? 

@cormacpayne What do we need to do for a breaking change release?

**Comment by cormacpayne** [comment]
LGTM once on demand build passes

http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1343/
